chips/nrf51/src/aes.rs

```diff
@@ -0,0 +1,256 @@
+//! AES128-CTR Driver
+//!
+//! Provides a simple driver for userspace applications to encrypt and decrypt messages
+//! using aes128-ctr mode on top of aes128-ecb
+//!
+//! At the moment are the initial counter always assigned to 0 but
+//! our idea is to have another allow call for that initial counter
+//! i.e. similar to the data(plaintext/cithertext)
+//!
+//! The buffer is also sliced in chips at the moment and some un-necessary
+//! static mut...
+//!
+//! FIXME:
+//!     - replace static mut with TakeCell or something similar
+//!     (I had problem to use because it can only be used one with take() )
+//!     - add support the enter initial counter value in userland
+//!     - maybe move some stuff to capsule instead
+//!
+//! Author: Niklas Adolfsson <niklasadolfsson1@gmail.com>
+//! Author: Fredrik Nilsson <frednils@student.chalmers.se>
+//! Date: March 16, 2017
+
+
+use chip;
+use core::cell::Cell;
+// use kernel::common::take_cell::TakeCell;
+use kernel::hil::symmetric_encryption::{SymmetricEncryptionDriver, Client};
+use nvic;
+use peripheral_interrupts::NvicIdx;
+use peripheral_registers::{AESECB_REGS, AESECB_BASE};
+
+#[deny(no_mangle_const_items)]
+
+static mut ECB_DATA: [u8; 48] = [0; 48];
+// key 0-15
+// cleartext 16-32
+// ciphertext 32-47
+
+static mut BUF: [u8; 128] = [0; 128];
+static mut DATA: [u8; 128] = [0; 128];
+static mut DMY: [u8; 16] = [0; 16];
+
+#[no_mangle]
+pub struct AesECB {
+    regs: *mut AESECB_REGS,
+    client: Cell<Option<&'static Client>>,
+    ctr: Cell<[u8; 16]>,
+    // data: TakeCell<'static, [u8]>,
+    remaining: Cell<u8>,
+    len: Cell<u8>,
+    offset: Cell<u8>,
+}
+
+pub static mut AESECB: AesECB = AesECB::new();
+
+impl AesECB {
+    const fn new() -> AesECB {
+        AesECB {
+            regs: AESECB_BASE as *mut AESECB_REGS,
+            client: Cell::new(None),
+            ctr: Cell::new([0; 16]),
+            // data: TakeCell::empty(),
+            remaining: Cell::new(0),
+            len: Cell::new(0),
+            offset: Cell::new(0),
+        }
+    }
+
+    pub fn ecb_init(&self) {
+        let regs = unsafe { &*self.regs };
+        unsafe {
+            regs.ECBDATAPTR.set((&ECB_DATA as *const u8) as u32);
+        }
+    }
+
+    // FIXME: should this be performed in constant time i.e. skip the break part and always loop 16
+    // times?
```

Comment:
I don't think this is an issue. For one, in general constant time operations are important for the cipher stage itself, not the stream. Second, this could only reveal the counter value for a particular block, which is not secret.